xla/service/gpu/gpu_compiler.cc

```diff
@@ -2003,34 +2010,45 @@ absl::StatusOr<GpuCompiler::BackendCompileResult> GpuCompiler::CompileAndLink(
             reinterpret_cast<const uint8_t*>(entry.binary().data());
         binaries_to_link.push_back(
             std::vector<uint8_t>(binary, binary + entry.binary().size()));
-        VLOG(5) << "Loaded " << name << ": " << entry.binary().size();
+        VLOG(5) << "Using " << name << " from cache: " << entry.binary().size();
         ++loaded_kernel_count;
       }
-      VLOG(2) << "Loaded " << loaded_kernel_count << " / "
-              << cache.entries_size() << " cached kernels.";
-    } else {
-      auto entries = cache.mutable_entries();
-      for (const auto& [name, binary] : binaries_to_cache) {
-        auto it = entries->find(name);
-        if (it == entries->end()) {
-          continue;
+      VLOG(2) << "Using " << loaded_kernel_count << " / "
+              << current_cache.entries_size() << " cached kernels.";
+    }
+    if (!binaries_to_cache.empty()) {
+      CompilationCacheProto disk_cache;
+      if (cache_file_exists) {
+        // Load the unmodified existing cache file to augment it.
```

Comment:
I think that now the code has the following overall logic:
1. load full cache in compile_module_to_llvm_ir.cc/LoadCache
2. filter out the needed kernels when doing `results.kernel_compilation_cache =
        ir_emitter_context.kernel_cache().Export();`
3. read the full cache again, add the new results to it, and write it out in gpu_compiler.cc

Would it be possible to simplify this flow or at least try to move all the cache reading and writing functions to one file, maybe kernel_reuse_cache.cc?
(e.g. why filter kernels if we load the discarded ones again? But maybe it makes sense from some perspective)
Now the logic of the caching is a bit hard to follow for me, because we have some kernel caching logic "here and there".

(This could be done in a separate CL potentially, if that makes sense.)

Reply:
Filtering is done to avoid linking unused kernels potentially present in the cache into the current executable. Another complexity for the overall flow is that at present not all compiled kernels are cacheable because of how llvm::SplitModule() works, I'll submit a patch there making it more convenient.
I'll draw a diagram and see what can be simplified.